src/ripple/app/misc/impl/ValidatorSite.cpp

```diff
@@ -165,153 +224,224 @@ ValidatorSite::onTimer (
-            }
-            else if (ListDisposition::invalid == disp)
-            {
-                JLOG (j_.warn()) <<
-                    "Invalid validator list from " <<
-                    sites_[siteIdx].uri;
-            }
-            else if (ListDisposition::unsupported_version == disp)
+void
+ValidatorSite::onSiteFetch(
+    boost::system::error_code const& ec,
+    detail::response_type&& res,
+    std::size_t siteIdx)
+{
+    Site::ResourcePtr newLocation;
+    bool shouldRetry = false;
+    {
+        std::lock_guard <std::mutex> lock_sites{sites_mutex_};
+        try
+        {
+            if (ec)
             {
                 JLOG (j_.warn()) <<
-                    "Unsupported version validator list from " <<
-                    sites_[siteIdx].uri;
+                    "Problem retrieving from " <<
+                    sites_[siteIdx].activeResource->uri <<
+                    " " <<
+                    ec.value() <<
+                    ":" <<
+                    ec.message();
+                shouldRetry = true;
+                throw std::runtime_error{"fetch error"};
             }
             else
             {
-                BOOST_ASSERT(false);
-            }
-
-            if (body.isMember ("refresh_interval") &&
-                body["refresh_interval"].isNumeric ())
-            {
-                sites_[siteIdx].refreshInterval =
-                    std::chrono::minutes{body["refresh_interval"].asUInt ()};
+                using namespace boost::beast::http;
+                if (res.result() == status::ok)
+                {
+                    parseJsonResponse(res, siteIdx, lock_sites);
+                }
+                else if (res.result() == status::moved_permanently  ||
+                         res.result() == status::permanent_redirect ||
+                         res.result() == status::found              ||
+                         res.result() == status::temporary_redirect)
+                {
+                    newLocation = processRedirect (res, siteIdx, lock_sites);
+                    // for perm redirects, also update our starting URI
+                    if (res.result() == status::moved_permanently ||
+                        res.result() == status::permanent_redirect)
+                    {
+                        sites_[siteIdx].startingResource = newLocation;
+                    }
+                }
+                else
+                {
+                    JLOG (j_.warn()) <<
+                        "Request for validator list at " <<
+                        sites_[siteIdx].activeResource->uri <<
+                        " returned bad status: " <<
+                        res.result_int();
+                    shouldRetry = true;
+                    throw std::runtime_error{"bad result code"};
+                }
+
+                if (newLocation)
+                {
+                    makeRequest(newLocation, siteIdx, lock_sites);
```

Comment:
Can throw with a secure scheme and adversely affect the fetching logic.

Reply:
in this case, the exception will be handled and I guess we'll retry again at the regular update interval. I think the real question for both calls to `makeRequest`, what should we do if `registerSLLCerts` is failing? Is retrying really going to help? Seems like a systemic issue, but of course we shouldn't crash.